source/opt/dominator_tree.cpp

```diff
@@ -0,0 +1,395 @@
+        BasicBlock* Succ = GetSuccessorBasicBlock(successorID);
+        SuccList.push_back(Succ);
+        Predecessors[Succ].push_back(&BB);
+      });
+    }
+  }
+}
+
+}  // namespace
+
+namespace spvtools {
+namespace opt {
+
+bool DominatorTree::StrictlyDominates(uint32_t A, uint32_t B) const {
+  if (A == B) return false;
+  return Dominates(A, B);
+}
+
+bool DominatorTree::StrictlyDominates(const ir::BasicBlock* A,
+                                      const ir::BasicBlock* B) const {
+  return DominatorTree::StrictlyDominates(A->id(), B->id());
+}
+
+bool DominatorTree::Dominates(uint32_t A, uint32_t B) const {
+  // Check that both of the inputs are actual nodes.
+  auto aItr = Nodes.find(A);
+  auto bItr = Nodes.find(B);
+  if (aItr == Nodes.end() || bItr == Nodes.end()) return false;
+
+  // Node A dominates node B if they are the same.
+  if (A == B) return true;
+  const DominatorTreeNode* nodeA = &aItr->second;
+  const DominatorTreeNode* nodeB = &bItr->second;
+
+  if (nodeA->DepthFirstPreOrderIndex < nodeB->DepthFirstPreOrderIndex &&
+      nodeA->DepthFirstPostOrderIndex > nodeB->DepthFirstPostOrderIndex) {
+    return true;
+  }
+
+  return false;
+}
+
+bool DominatorTree::Dominates(const ir::BasicBlock* A,
+                              const ir::BasicBlock* B) const {
+  return Dominates(A->id(), B->id());
+}
+
+ir::BasicBlock* DominatorTree::ImmediateDominator(
+    const ir::BasicBlock* A) const {
+  return ImmediateDominator(A->id());
+}
+
+ir::BasicBlock* DominatorTree::ImmediateDominator(uint32_t A) const {
+  // Check that A is a valid node in the tree.
+  auto aItr = Nodes.find(A);
+  if (aItr == Nodes.end()) return nullptr;
+
+  const DominatorTreeNode* nodeA = &aItr->second;
+
+  if (nodeA->Parent == nullptr) {
+    return nullptr;
+  }
+
+  return nodeA->Parent->BB;
+}
+
+DominatorTreeNode* DominatorTree::GetOrInsertNode(ir::BasicBlock* BB) {
+  DominatorTreeNode* DTN = nullptr;
+
+  std::map<uint32_t, DominatorTreeNode>::iterator ItNode = Nodes.find(BB->id());
+  if (ItNode == Nodes.end()) {
+    DTN = &Nodes.emplace(std::make_pair(BB->id(), DominatorTreeNode{BB}))
+               .first->second;
+  } else
+    DTN = &ItNode->second;
+
+  return DTN;
+}
+
+void DominatorTree::GetDominatorEdges(
+    const ir::Function* F, ir::BasicBlock* DummyStartNode,
+    std::vector<std::pair<ir::BasicBlock*, ir::BasicBlock*>>& edges) {
+  // Each time the depth first traversal calls the postorder callback
+  // std::function we push that node into the postorder vector to create our
+  // postorder list.
+  std::vector<const ir::BasicBlock*> postorder;
+  auto postorder_function = [&](const ir::BasicBlock* b) {
+    postorder.push_back(b);
+  };
+
+  // CFA::CalculateDominators requires std::vector<ir::BasicBlock*>
+  // BB are derived from F, so we need to const cast it at some point
+  // no modification is made on F.
+  BasicBlockSuccessorHelper<ir::BasicBlock> helper{
```

Comment:
Please also run clang-format over the code.  This seems oddly formatted.

Reply:
I did, I just ran it again to check, no changes

Reply:
Thanks.  It's the initializer list that tricked me.